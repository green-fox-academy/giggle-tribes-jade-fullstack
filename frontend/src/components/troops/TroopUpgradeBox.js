import React from 'react';
import { connect } from 'react-redux';
import { Button } from '@material-ui/core';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import PropTypes from 'prop-types';

import './TroopButton.css';
import { upgradeTroopAction } from '../../actions/TroopsActions';
import { setErrorAction } from '../../actions/ErrorActions';

function TroopUpgradeBox({
  open,
  setOpen,
  level,
  troopAmount,
  kingdom,
  upgrade,
  resources,
  buildings,
  setError,
}) {
  const [amount, setAmount] = React.useState(0);

  const handleClose = () => {
    setOpen(false);
  };

  const upgradeTroopClickOn = () => {
    const troopUpgradeConditions = {
      upgradeCost: 10,
      goldAmount: resources.find(resource => resource.type === 'gold').amount,
      academyLevel:
        buildings.length > 0
          ? buildings.find(e => e.type === 'academy').level
          : 1,
    };

    if (
      troopUpgradeConditions.goldAmount >= troopUpgradeConditions.upgradeCost &&
      level < troopUpgradeConditions.academyLevel &&
      amount <= troopAmount
    ) {
      upgrade(kingdom, amount, level);
    } else if (
      troopUpgradeConditions.goldAmount < troopUpgradeConditions.upgradeCost
    ) {
      setError("You don't have enough money.");
    } else if (amount > troopAmount) {
      setError(
        `Amount was too much, you have ${troopAmount} troops in that troop level`
      );
    } else {
      setError('Upgrade is not allowed, academy level too low');
    }
    handleClose();
  };

  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">Troop Upgrade</DialogTitle>
      <DialogContent>
        <DialogContentText>
          Please provide amount of troops to be upgraded and click upgrade.
        </DialogContentText>
        <TextField
          autoFocus
          margin="dense"
          id="name"
          label="Amount"
          type="number"
          InputProps={{ inputProps: { min: 0 } }}
          onChange={e => {
            setAmount(e.target.value);
          }}
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          Cancel
        </Button>
        <Button
          className="TroopUpgradeBTN"
          onClick={upgradeTroopClickOn}
          color="primary"
        >
          Upgrade
        </Button>
      </DialogActions>
    </Dialog>
  );
}

TroopUpgradeBox.propTypes = {
  open: PropTypes.bool.isRequired,
  setOpen: PropTypes.func.isRequired,
  level: PropTypes.number.isRequired,
  troopAmount: PropTypes.number.isRequired,
  kingdom: PropTypes.string.isRequired,
  troops: PropTypes.array.isRequired,
  resources: PropTypes.array.isRequired,
  buildings: PropTypes.array.isRequired,
  setError: PropTypes.func.isRequired,
};

const mapStateToProps = ({ kingdom, troops, resources, buildings }) => {
  return { kingdom, troops, resources, buildings };
};

const mapDispatchToProps = dispatch => {
  return {
    upgrade: (kingdomId, amount) => {
      dispatch(upgradeTroopAction(kingdomId, amount));
    },
    setError: error => {
      dispatch(setErrorAction(error));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TroopUpgradeBox);
