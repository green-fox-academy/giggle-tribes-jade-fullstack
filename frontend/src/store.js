import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import { userReducer } from './reducers/userReducer';
import { kingdomReducer } from './reducers/kingdomReducer';
import { buildingsReducer } from './reducers/buildingsReducer';
import { resourcesReducer } from './reducers/resourcesReducer';
import { troopsReducer } from './reducers/troopsReducer';
import { tokenReducer } from './reducers/tokenReducer';
import { errorReducer } from './reducers/errorReducer';

function root(state = {}, action) {
  return {
    user: userReducer(state.user, action),
    kingdom: kingdomReducer(state.kingdom, action),
    buildings: buildingsReducer(state.buildings, action),
    resources: resourcesReducer(state.resources, action),
    troops: troopsReducer(state.troops, action),
    token: tokenReducer(state.token, action),
    error: errorReducer(state.error, action),
  };
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(
  root,
  /* preloadedState, */ composeEnhancers(applyMiddleware(thunk))
);

export { store };
