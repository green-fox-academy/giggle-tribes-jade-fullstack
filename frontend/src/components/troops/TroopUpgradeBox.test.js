import React from 'react';
import { Provider } from 'react-redux';
import { render, unmountComponentAtNode } from 'react-dom';
import { act, Simulate } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';

import TroopUpgradeBox from './TroopUpgradeBox';
import { store } from '../../store';

global.fetch = jest.fn(async () => {
  return await Promise.resolve({
    json: () =>
      Promise.resolve({
        troops: [
          {
            id: 1,
          },
          {
            id: 2,
          },
        ],
      }),
  });
});

let container;

beforeEach(() => {
  fetch.mockClear();
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('TroopUpgradeBox tests', () => {
  it('should dispatch an action when clicked', async () => {
    const mockStore = configureStore([]);
    const mockedStore = mockStore({
      kingdom: '1',
      buildings: [{ type: 'academy', level: 2 }],
      resources: [
        {
          type: 'food',
          amount: 57024,
          generation: 1,
          updatedAt: '2020-08-12T17:47:32.000Z',
        },
        {
          type: 'gold',
          amount: 57024,
          generation: 1,
          updatedAt: '2020-08-12T17:47:32.000Z',
        },
      ],
      troops: [],
    });
    mockedStore.dispatch = jest.fn();
    await act(async () => {
      render(
        <Provider store={mockedStore}>
          <TroopUpgradeBox
            open={true}
            setOpen={() => {}}
            level={1}
            troopAmount={10}
          />
        </Provider>,
        container
      );
    });

    const button = document.getElementsByClassName('TroopUpgradeBTN')[0];
    Simulate.click(button);
    expect(mockedStore.dispatch).toHaveBeenCalledTimes(1);
  });

  it('should dispatch an action when clicked, not enough money', async () => {
    const mockStore = configureStore([]);
    const mockedStore = mockStore({
      kingdom: '1',
      buildings: [{ type: 'academy', level: 2 }],
      resources: [
        {
          type: 'food',
          amount: 57024,
          generation: 1,
          updatedAt: '2020-08-12T17:47:32.000Z',
        },
        {
          type: 'gold',
          amount: 9,
          generation: 1,
          updatedAt: '2020-08-12T17:47:32.000Z',
        },
      ],
      troops: [],
    });
    mockedStore.dispatch = jest.fn();
    await act(async () => {
      render(
        <Provider store={mockedStore}>
          <TroopUpgradeBox
            open={true}
            setOpen={() => {}}
            level={1}
            troopAmount={10}
          />
        </Provider>,
        container
      );
    });
    const button = document.getElementsByClassName('TroopUpgradeBTN')[0];
    Simulate.click(button);
    expect(mockedStore.dispatch).toHaveBeenCalledTimes(1);
  });

  it('should dispatch an action when clicked, academy level is too low', async () => {
    const mockStore = configureStore([]);
    const mockedStore = mockStore({
      kingdom: '1',
      buildings: [{ type: 'academy', level: 1 }],
      resources: [
        {
          type: 'food',
          amount: 57024,
          generation: 1,
          updatedAt: '2020-08-12T17:47:32.000Z',
        },
        {
          type: 'gold',
          amount: 57024,
          generation: 1,
          updatedAt: '2020-08-12T17:47:32.000Z',
        },
      ],
      troops: [],
    });
    mockedStore.dispatch = jest.fn();
    await act(async () => {
      render(
        <Provider store={mockedStore}>
          <TroopUpgradeBox
            open={true}
            setOpen={() => {}}
            level={1}
            troopAmount={10}
          />
        </Provider>,
        container
      );
    });
    const button = document.getElementsByClassName('TroopUpgradeBTN')[0];
    Simulate.click(button);
    expect(mockedStore.dispatch).toHaveBeenCalledTimes(1);
  });

  it('matches snapeshot', async () => {
    Storage.prototype.getItem = jest.fn(key => {
      return 'dummy';
    });
    await act(async () => {
      render(
        <Provider store={store}>
          <TroopUpgradeBox
            open={true}
            setOpen={() => {}}
            level={1}
            troopAmount={10}
          />
        </Provider>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
  });
});
