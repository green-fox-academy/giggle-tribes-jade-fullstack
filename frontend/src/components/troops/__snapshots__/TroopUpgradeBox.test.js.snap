// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TroopUpgradeBox tests matches snapeshot 1`] = `""`;
